Fix noms-sync surprising quantity

ehalpern opened the pull request:
In datas.Pull, count the number of chunks to be written and serialize a sample to estimate number of bytes written. 

toward: #1998

Commit:
Use sampling for a better bytes-written estimate for noms sync

toward: #1998

aboodman commented on go/datas/pull.go:
```diff
@@ -111,12 +118,22 @@ func Pull(srcDB, sinkDB Database, sourceRef, sinkHeadRef types.Ref, concurrency
+					// Refs received on this channel represent chunks to be written to the sink. To
+					// estimate bytes written, take a sample of these chunks and measure their serialized size.
+					// Since each sample requires reading a chunk from the source, so it's important to manage
+					// the sampling rate so that sampling slows things down.
```
This last sentence doesn't parse. Remove first "is"? Also do you mean "so that sample *doesn't* slow things down?"

ehalpern replied:
Yes. Should read: 

// Since each sample requires reading a chunk from the source, it's important to manage
// the sampling rate to avoid slow down

Commit:
Fix comment typos

aboodman commented on go/datas/pull.go:
```diff
@@ -80,23 +83,27 @@ func Pull(srcDB, sinkDB Database, sourceRef, sinkHeadRef types.Ref, concurrency
-		doneCount, knownCount, doneBytes = doneCount+moreDone, knownCount+moreKnown, doneBytes+moreBytes
-		progressCh <- PullProgress{doneCount, knownCount + uint64(srcQ.Len()), doneBytes}
+		doneCount, knownCount, doneBytes, approxBytesWritten = doneCount+moreDone, knownCount+moreKnown, doneBytes+moreBytesRead, approxBytesWritten+moreApproxBytesWritten
+		progressCh <- PullProgress{doneCount, knownCount + uint64(srcQ.Len()), doneBytes, approxBytesWritten /*sinkDB.validatingBatchStore().(*localBatchStore).BytesWritten()*/}
```
Nit: unbalanced whitespace: ` approxBytesWritten /*sinkDB.validatingBatchStore...`

aboodman commented on go/datas/pull.go:
```diff
@@ -111,12 +118,22 @@ func Pull(srcDB, sinkDB Database, sourceRef, sinkHeadRef types.Ref, concurrency
+					// Refs received on this channel represent chunks to be written to the sink. To
+					// estimate bytes written, take a sample of these chunks and measure their serialized size.
+					// Since each sample requires reading a chunk from the source, it's important to manage
+					// the sampling rate to avoid slow down.
```
Nit: typo: `slowdown`

aboodman commented on go/datas/pull.go:
```diff
@@ -265,3 +282,35 @@ func traverseCommon(comRef, sinkHead types.Ref, db Database) traverseResult {
+	increment := math.Min(1000, math.Max(3.0, num / denom))
+	// choose the increment randomly with increment as the midpoint
+	randIncrement := uint64(rand.Int63n(int64(increment + increment / 2))) + 1
+	//fmt.Printf("next increment from %d: rand(%d) = %d\n", curIdx, uint64(increment), randIncrement)
```
Remove deadcode

arv commented:
jenkins: test this please

arv commented:
jenkins: test this please

Commit:
Remove dead code and fix typos

toward: #2531

rafael-atticlabs commented on go/datas/pull.go:
```diff
@@ -5,16 +5,19 @@
 
 type PullProgress struct {
-	DoneCount, KnownCount, DoneBytes uint64
+	DoneCount, KnownCount, DoneBytes, ApproxWrittenBytes uint64
```
no one is using `DoneBytes` anymore, it looks like. Remove?

rafael-atticlabs commented:
In general, this looks good. My main suggestion is to move the sampling into the `traverseSource` so as to avoid additional reads.

Commit:
Improve sync bytes written estimation by sampling in traverseSource

These changes based on @rafael-atticlabs suggestions reduce the cost of sampling
and simplify the code.

- Moved sampling into traverseSource where chunk has already been retrieve
  from source

- Simplified sampling logic since sampling is now cheap enough to always
  sample at 10%

- Tweak what is sampled to improve accuracy. This may point to an
  underlying bug or lack of understanding on my part. See
  pull.estBytesWritten

toward: #1998

Commit:
Remove DoneBytes from PullProgress

Also add notes comparing sampling rates

toward: #2531

Commit:
Confirm that overestimate of data written is consistent with leveldb stats

Did more investigation to determine whether the ~7-15% descrepency between
estimated bytes written and byte size of level-db directory is a concern.
One would expect the actual size on disk to be larger than estimate due to
leveldb overhead, but we're seeing the opposite. I've confirmed that
there no double counting of references. I've also confirmed that
the estimate consistent with stats reported using noms sync --lb-dump-stats.
This indicates that the issue is somewhere in the ldb driver and not a
noms bug.

fixes: #1998

Commit:
Adjust pull_test to work with approxBytesWritten

ehalpern commented:
PTAL @rafael-atticlabs. I've another issue for the overestimation: #2567

rafael-atticlabs commented on go/datas/pull.go:
```diff
@@ -64,7 +67,13 @@ func Pull(srcDB, sinkDB Database, sourceRef, sinkHeadRef types.Ref, concurrency
+				// the serialized, compressed bytes of each chunk, we take a 10% sample.
+				// There's no immediately observable performance benefit to sampling here, but there's
+				// also no appreciable loss in accuracy, so we'll keep it around.
+					takeSample := rand.Float64() < .10
```
let's make the .10 a const and put it above.

rafael-atticlabs commented:
lgtm w/ nits. awesome. thanks for digging deep on this one.

Commit:
Use constant for bytesWrittenSampleRate

toward: #1998
